// @remotion/cli/dist/parsed-cli.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.quietFlagProvided = exports.parsedCli = exports.BooleanFlags = void 0;
const client_1 = require("@remotion/renderer/client");
const minimist_1 = __importDefault(require("minimist"));
exports.BooleanFlags = [
    'overwrite',
    'force',
    'sequence',
    'help',
    'quiet',
    'q',
    'muted',
    client_1.BrowserSafeApis.options.enforceAudioOption.cliFlag,
    // Lambda flags
    'force',
    'disable-chunk-optimization',
    'save-browser-logs',
    'disable-cloudwatch',
    'enable-lambda-insights',
    'yes',
    'y',
    'disable-web-security',
    client_1.BrowserSafeApis.options.darkModeOption.cliFlag,
    'ignore-certificate-errors',
    'disable-headless',
    'disable-keyboard-shortcuts',
    'default-only',
    'no-open',
    'ipv4',
    client_1.BrowserSafeApis.options.beepOnFinishOption.cliFlag,
    client_1.BrowserSafeApis.options.disableGitSourceOption.cliFlag,
    client_1.BrowserSafeApis.options.disallowParallelEncodingOption.cliFlag,
    client_1.BrowserSafeApis.options.forSeamlessAacConcatenationOption.cliFlag,
    'repro',
    'compatible-only',
    'force-path-style',
    'onlyAllocateCpuDuringRequestProcessing',
];
exports.parsedCli = (0, minimist_1.default)(process.argv.slice(2), {
    boolean: exports.BooleanFlags,
    default: {
        overwrite: true,
        muted: null,
    },
});
const quietFlagProvided = () => exports.parsedCli.quiet || exports.parsedCli.q;
exports.quietFlagProvided = quietFlagProvided;
