// @remotion/renderer/dist/options/is-production-key.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isProductionKeyOption = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const cliFlag = 'is-production-key';
let currentIsProductionKey = null;
exports.isProductionKeyOption = {
    name: 'Is Production Key',
    cliFlag,
    description: () => (jsx_runtime_1.jsx(jsx_runtime_1.Fragment, { children: "Pass \"true\" if you are rendering in production. All the production renders are considered as billable renders on the Pro platform." })),
    ssrName: 'isProductionKey',
    docLink: 'https://www.remotion.dev/docs/licensing',
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== undefined) {
            return {
                source: 'cli',
                value: commandLine[cliFlag],
            };
        }
        if (currentIsProductionKey !== null) {
            return {
                source: 'config',
                value: currentIsProductionKey.toString(),
            };
        }
        return {
            source: 'default',
            value: null,
        };
    },
    setConfig: (value) => {
        currentIsProductionKey = value === 'true';
    },
    type: null,
};
